cli/internal/runsummary/spaces.go

```diff
@@ -1,15 +1,259 @@
+		runCreated: make(chan struct{}, 1),
+		run:        &spaceRun{},
+	}
+}
+
+// Start receiving and processing requests in 8 goroutines
+// There is an additional marker (protected by a mutex) that indicates
+// when the first request is done. All other requests are blocked on that one.
+// This first request is the POST /run request. We need to block on it because
+// the response contains the run ID from the server, which we need to construct the
+// URLs of subsequent requests.
+func (c *spacesClient) start() {
+	if !c.enabled {
+		return
+	}
+
+	// Start an immediately invoked go routine that listens for requests coming in from a channel
+	pending := []*spaceRequest{}
+	firstRequestStarted := false
+
+	// Create a labeled statement so we can break out of the for loop more easily
+
+	// Setup a for loop that goes infinitely until we break out of it
+FirstRequest:
+	for {
+		// A select statement that can listen for messages from multiple channels
+		select {
+		// listen for new requests coming in
+		case req := <-c.requests:
+			// Make the first request right away in a goroutine,
+			// queue all other requests. When the first request is done,
+			// we'll get a message on the other channel and break out of this loop
+			if !firstRequestStarted {
+				firstRequestStarted = true
+				go c.dequeueRequest(req)
+			} else {
+				pending = append(pending, req)
+			}
+			// Wait for c.runCreated channel to be closed and:
+		case <-c.runCreated:
+			// 1. flush pending requests
+			for _, req := range pending {
+				go c.dequeueRequest(req)
+			}
+
+			// 2. break out of the forever loop.
+			break FirstRequest
+		}
+	}
+
+	// and then continue listening for more requests as they come in until the channel is closed
+	for req := range c.requests {
+		go c.dequeueRequest(req)
+	}
+}
+
+func (c *spacesClient) makeRequest(req *spaceRequest) {
+	if !c.enabled {
+		return
+	}
+
+	// The runID is required for POST task requests and PATCH run request URLS,
+	// so we have to construct these URLs lazily with a `makeURL` affordance.
+	//
+	// We are assuming that if makeURL is defined, this is NOT the first request.
+	// This is not a great assumption, and will fail if our endpoint URLs change later.
+	//
+	// Secondly, if makeURL _is_ defined, we call it, and if there are any errors, we exit early.
+	// We are doing this check before any of the other basic checks (e.g. the existence of a spaceID)
+	// becaus in the case the repo is not linked to a space, we don't want to print those errors
+	// for every request that fails. On the other hand, if that POST /run request fails, and N
+	// requests fail after that as a consequence, it is ok to print all of those errors.
+	if req.makeURL != nil {
+		if err := req.makeURL(req, c.run); err != nil {
+			c.errors = append(c.errors, err)
+			return
+		}
+	}
+
+	if c.spaceID == "" {
+		c.errors = append(c.errors, req.error("No spaceID found"))
+		return
+	}
+
+	if !c.api.IsLinked() {
```

Comment:
Maybe this ought to be checked earlier in the process? Like, maybe consider spaces disabled if the api client is not linked?